services/wallet/chain/client.go

```diff
@@ -46,63 +55,93 @@ func NewClients(rpc *rpc.Client, chainIDs []uint64) (res []*Client, err error) {
 	return res, nil
 }
 
+func (cc *Client) toggleIsConnected(err error) {
+	if err != nil {
+		cc.IsConnected = false
+	} else {
+		cc.IsConnected = true
+	}
+}
+
 func (cc *Client) ToBigInt() *big.Int {
 	return big.NewInt(int64(cc.ChainID))
 }
 
 func (cc *Client) HeaderByHash(ctx context.Context, hash common.Hash) (*types.Header, error) {
 	rpcstats.CountCall("eth_getBlockByHash")
-	return cc.eth.HeaderByHash(ctx, hash)
+	resp, err := cc.eth.HeaderByHash(ctx, hash)
+	defer cc.toggleIsConnected(err)
```

Comment:
Hmm I'd like avoiding repetition by moving this to a single location like you did in `doQuery`. I'd need to do some reading to know how to do this in go, but I'm guessing passing a lambda instead of a URL. Not sure it ends up being tidier, though.

Reply:
I might be missing something but I don't see how i could avoid to call the `defer` here, to me it is similar than countCall being on all function. Happy to change to you suggestion, if you can help me how to